lsp/zk.lua

```diff
@@ -1,74 +1,117 @@
+  opts = vim.tbl_extend('keep', { select = { 'path', 'title' } }, opts or {})
+  client:exec_cmd({
+    title = 'ZkList',
+    command = 'zk.list',
+    arguments = { vim.api.nvim_buf_get_name(bufnr), opts },
+  }, { bufnr = bufnr }, function(err, result)
+    if err ~= nil then
+      vim.api.nvim_echo({ { 'zk.list error\n' }, { vim.inspect(err) } }, true, {})
+      return
     end
-  end
+    if result == nil then
+      return
+    end
+
+    local paths = vim.tbl_map(function(item)
+      return item.path
+    end, result)
+    local titles = vim.tbl_map(function(item)
+      return item.title
+    end, result)
+    vim.ui.select(titles, {}, function(item, idx)
+      if item ~= nil then
+        action(paths[idx], item)
+      end
+    end)
+  end)
 end
 
 ---@type vim.lsp.Config
 return {
   cmd = { 'zk', 'lsp' },
   filetypes = { 'markdown' },
   root_markers = { '.zk' },
-  ---@param bufnr integer
-  ---@param client vim.lsp.Client
+  workspace_required = true,
   on_attach = function(client, bufnr)
     vim.api.nvim_buf_create_user_command(bufnr, 'LspZkIndex', function()
       client:exec_cmd({
         title = 'ZkIndex',
         command = 'zk.index',
         arguments = { vim.api.nvim_buf_get_name(bufnr) },
-      }, { bufnr = bufnr })
-    end, {
-      desc = 'ZkIndex',
-    })
+      }, { bufnr = bufnr }, function(err, result)
+        if err ~= nil then
+          vim.api.nvim_echo({ { 'zk.index error\n' }, { vim.inspect(err) } }, true, {})
+          return
+        end
+        if result ~= nil then
+          vim.api.nvim_echo({ { vim.inspect(result) } }, false, {})
+        end
+      end)
+    end, { desc = 'ZkIndex' })
 
     vim.api.nvim_buf_create_user_command(bufnr, 'LspZkList', function()
-      local bufpath = vim.api.nvim_buf_get_name(bufnr)
-      local root = find_zk_root(bufpath)
+      zk_list(client, bufnr, {}, function(path)
+        vim.cmd('edit ' .. path)
+      end)
+    end, { desc = 'ZkList' })
 
+    vim.api.nvim_buf_create_user_command(bufnr, 'LspZkTagList', function()
       client:exec_cmd({
-        title = 'ZkList',
-        command = 'zk.list',
-        arguments = { root, { select = { 'path' } } },
-      }, { bufnr = bufnr }, function(_err, result)
-        if not result then
+        title = 'ZkTagList',
+        command = 'zk.tag.list',
+        arguments = { vim.api.nvim_buf_get_name(bufnr) },
+      }, { bufnr = bufnr }, function(err, result)
+        if err ~= nil then
+          vim.api.nvim_echo({ { 'zk.tag.list error\n' }, { vim.inspect(err) } }, true, {})
+          return
+        end
+        if result == nil then
           return
         end
-        local paths = vim.tbl_map(function(item)
-          return item.path
+
+        local tags = vim.tbl_map(function(item)
+          return item.name
         end, result)
-        vim.ui.select(paths, {}, function(choice)
-          vim.cmd('edit ' .. choice)
+        vim.ui.select(tags, {}, function(item)
```

Comment:
use `{format_item=...}` to avoid the `tbl_map`, as shown in my other comment.

the reason this matters is because people copy what other configs are doing, so doing it right in this config helps avoid problems in the future.